#166840928 Users can Follow One Another

tolumide-ng opened the pull request:
#### Description
Users can follow or unfollow one another  via **/api/v1/follow** or **/api/v1/unfollow** respectively
so far they both are registered users of the errorSwag application 

#### Type of Change
- [x] New feature (Non-breaking change which adds functionality)
- [x] Bug fix (Non-breaking change which adds functionality)


#### How Has This Been Tested?
- [x] Integration Testing


#### Pivotal Stories
[#166840928](https://www.pivotaltracker.com/story/show/166840928)

segunolalive commented on src/controllers/user.controller.js:
```diff
@@ -172,6 +172,83 @@ class UserController {
+        followeeId
+      };
+
+      const followExist = await BaseRepository.findOneByField(
```
Please change `followExist` to something `follower` because `followExist` suggests a boolean value.

tolumide-ng replied:
Change effected

segunolalive commented on src/middleware/users.middleware.js:
```diff
@@ -0,0 +1,34 @@
+  next();
+};
+
+export const isSelf = async (req, res, next) => {
```
I like this. ❤️

tolumide-ng replied:
Thank you 👍

tohbay replied:
The `isSelf` variable declaration does clearly say what the variable is representing.

It will be better the variable declaration is self-descriptive since it is a middleware function that someone else on the team may need to make use of.

tolumide-ng replied:
> The `isSelf` variable declaration does clearly say what the variable is representing.
> 
> It will be better the variable declaration is self-descriptive since it is a middleware function that someone else on the team may need to make use of.

Thank you for your response @tobechukwuobitube , could you suggest a better name that I could use?

segunolalive commented on src/test/user.spec.js:
```diff
@@ -192,3 +200,139 @@ describe('Test user login, signup and account verification', () => {
+      .patch(`${users}/follow`)
+      .set('token', token)
+      .send({ followeeId: secondUser.id });
+    expect(res.status).to.equal(200);
```
Test that the second user actually got a new follwer whose Id === firstUser.id

tolumide-ng replied:
Changes effected thank you

segunolalive commented on src/test/user.spec.js:
```diff
@@ -192,3 +200,139 @@ describe('Test user login, signup and account verification', () => {
+    expect(res.status).to.equal(200);
+  });
+
+  it('should not be able to follow a user twice at the same time', async () => {
```
end this statement at `twice`

tolumide-ng replied:
Thank you for your comment, I have effected the changes

segunolalive commented on src/test/user.spec.js:
```diff
@@ -192,3 +200,139 @@ describe('Test user login, signup and account verification', () => {
+    );
+  });
+
+  it('user should not be able to themselve', async () => {
```
... to follow themselves

tolumide-ng replied:
Thank you for your comment, I have rephrased the sentence

segunolalive commented on src/test/user.spec.js:
```diff
@@ -192,3 +200,139 @@ describe('Test user login, signup and account verification', () => {
+  it('should not be able to follow a user twice at the same time', async () => {
+    const firstUser = await createUser(await getUser());
+    const secondUser = await createUser(await getUser());
+    await baseRepository.create(db.Follower, {
```
use the public API instead.

Send the request `twice`. Assert that the user was followed once initially but the second request failed.

segunolalive commented on src/test/user.spec.js:
```diff
@@ -192,3 +200,139 @@ describe('Test user login, signup and account verification', () => {
+      .set('token', token)
+      .send({ followeeId: firstUser.id });
+    expect(res.status).to.equal(400);
+    expect(res.body.message).to.equal(`You cannot follow or unfollow yourself`);
```
Assert that the follower count hasn't increased and followers doesn't include the user's id

tolumide-ng replied:
Thank you for your comment, Changes effected

segunolalive commented on src/test/user.spec.js:
```diff
@@ -192,3 +200,139 @@ describe('Test user login, signup and account verification', () => {
+    expect(res.body.message).to.equal(`You cannot follow or unfollow yourself`);
+  });
+
+  it('throw validation error if id of the user is not supplied', async () => {
```
throws

segunolalive commented on src/test/user.spec.js:
```diff
@@ -192,3 +200,139 @@ describe('Test user login, signup and account verification', () => {
+
+  it('should not be able to follow a non-existing user', async () => {
+    const firstUser = await createUser(await getUser());
+    await createUser(await getUser());
```
this is unused

segunolalive commented on src/test/user.spec.js:
```diff
@@ -192,3 +200,139 @@ describe('Test user login, signup and account verification', () => {
+
+  it('throw validation error if id of the user supplied is not a number', async () => {
+    const firstUser = await createUser(await getUser());
+    await createUser(await getUser());
```
not used

segunolalive commented on src/test/user.spec.js:
```diff
@@ -192,3 +200,139 @@ describe('Test user login, signup and account verification', () => {
+    const res = await server()
+      .patch(`${users}/follow`)
+      .set('token', token);
+    expect(res.status).to.equal(400);
```
test that the right message is sent.

segunolalive reviewed (commented):
Good Job. 👍 

See comments below.

Endeavour to update parts of the PR I didn't comment on but share similar behaviour with the ones I commented on.

olorunwalawrence commented on package.json:
```diff
@@ -4,8 +4,8 @@
     "main": "index.js",
     "scripts": {
-        "pretest": "sequelize db:migrate:undo:all && sequelize db:migrate  && sequelize db:seed:all",
-        "test": "nyc --require @babel/register --require @babel/polyfill mocha src/test/**.spec.js --timeout 20000 --exit",
```
why do you have two test script?, I would that you remove one of the test script in the package.json.

olorunwalawrence commented on src/controllers/user.controller.js:
```diff
@@ -1,10 +1,10 @@
+import responseGenerator from "../helpers/responseGenerator";
+import utility from "../helpers/utils";
 
-import db from '../database/models';
```
Double import of db file, kindly fix it.

tolumide-ng replied:
Thank you

olorunwalawrence commented on src/controllers/user.controller.js:
```diff
@@ -162,16 +159,93 @@ class UserController {
+        db.Follower,
+        details
+      );
+      if (!follower) {
```
it would be better to check if the follower returns true then send a success response then check if the user returns false or null return error message

tolumide-ng replied:
Hello @olorunwalawrence , Thank you for your response😊I do not understand this review (I would appreciate some more clarity) but I would assume you are referring to confirming the existence of the user to be followed/unfollowed which as already been done on a middleware before this (to ensure reusability for others who might need such functionality).
Please check that here: https://github.com/andela/ah-kifaru-backend/blob/ft-users-follow-oneAnother-166840928/src/middleware/users.middleware.js

olorunwalawrence commented on src/controllers/user.controller.js:
```diff
@@ -162,16 +159,93 @@ class UserController {
+  static async followUser(req, res) {
+    try {
+      //  Please inform team to change the id of the decrypted user to id when been parsed with req object
+      const { followeeId } = req.body;
```
I think the followee id should come from req.params and not from the request body; you would not want the user to typed in their userId

tolumide-ng replied:
Thank you for your insight, I will implement the changes.

timi-codes commented on src/middleware/users.middleware.js:
```diff
@@ -0,0 +1,34 @@
+    return responseGenerator.sendError(
+      res,
+      400,
+      `You cannot follow or unfollow yourself`
```
Good initiative!

segunolalive commented on src/test/user.spec.js:
```diff
@@ -193,3 +200,226 @@ describe('Test user login, signup and account verification', () => {
   });
 });
+
+describe('USER CAN FOLLOW OTHER USERS', () => {
```
Set the description to the route (API) being tested

timi-codes replied:
Include the endpoint/routes you are testing in the description

segunolalive commented on src/routes/v1/index.js:
```diff
@@ -3,5 +3,6 @@ import userRoutes from './user.route';
 
 export default app => {
   app.use('/auth', authRoutes);
+  app.use('/api/v1/user', userRoutes);
```
This should be plural `/api/v1/users`. It represents the `users` collection

tolumide-ng replied:
changes made. Thank you

timi-codes commented on src/test/user.spec.js:
```diff
@@ -193,3 +200,226 @@ describe('Test user login, signup and account verification', () => {
+    const firstUser = await createUser();
+    const secondUser = await createUser();
+    await createUser();
+    const token = jwt.sign(firstUser, process.env.JWT_SECRET);
```
This method exists in already as a helper function. Kindly use that instead

tolumide-ng replied:
I have removed it now, thank you

segunolalive commented on src/test/user.spec.js:
```diff
@@ -1,10 +1,17 @@
+import db from '../database/models';
+import baseRepository from '../repository/base.repository';
+
+const users = '/api/v1/user';
```
use USER_API, USER_ENDPOINT ?? 🤷‍♂

timi-codes commented on src/test/user.spec.js:
```diff
@@ -193,3 +200,226 @@ describe('Test user login, signup and account verification', () => {
+    const res = await server()
+      .patch(`${users}/follow/${secondUser.id}`)
+      .set('token', token);
+    // .send({ followeeId: secondUser.id });
```
Remove the comment line of code

tolumide-ng replied:
Changes effected, thank you

segunolalive commented on src/test/user.spec.js:
```diff
@@ -193,3 +200,226 @@ describe('Test user login, signup and account verification', () => {
+    await db.User.destroy({ cascade: true, truncate: true });
+  });
+
+  it('should be able to follow other users', async () => {
```
test case description should read as continuation of test block description

tolumide-ng replied:
changes effected, thank you

segunolalive commented on src/test/user.spec.js:
```diff
@@ -193,3 +200,226 @@ describe('Test user login, signup and account verification', () => {
+    const res = await server()
+      .patch(`${users}/follow/abc`)
+      .set('token', token);
+    // .send({ followeeId: '' });
```
remove commented/unused code.

tolumide-ng replied:
Changes effected, thank you

timi-codes reviewed (changes requested):
Well done!
I left some comments and suggestions below.

segunolalive reviewed (changes requested):
👍

segunolalive commented on src/test/user.spec.js:
```diff
@@ -277,3 +287,226 @@ describe('GET api/v1/users', () => {
+
+  it('should not be able to follow itself', async () => {
+    const firstUser = await createUser();
+    await createUser();
```
Remove this line, please.

tolumide-ng replied:
Changes effected

segunolalive commented on src/test/utils/helpers.js:
```diff
@@ -0,0 +1,23 @@
+  return plain;
+};
+
+// export const jwtSigner = async payload => {
```
Remove dead code.

tolumide-ng replied:
changes made, thank you

segunolalive reviewed (changes requested):
Looking good. Would merge after the changes

tolumide-ng committed:
feat(users): Users can follows eachother

- Installs faker to simulate users for testing
- Users can follow one another
- Creates findOrCreate method in basrepository
- Validates user input
- Modifies .eslintrc file
- Ensures cannot follow themselve

[Delivers #166840920]
